### PR TITLE
Comedian neural network research request

### PR DESCRIPTION
**Pull request by iaroslav-ai**
Create a deep neural network capable of generating funny jokes.

### CODE REVIEW

**Comment by ilyasu123** [comment]
Scraping:  add a comment in the instructions to obey robots.txt or whatever request that the website has with respect to scraping.  I heard that reddit has exposed a few terabytes of reddit comments for download -- you should advice to get that.  In fact, I believe that your problem will have a much greater chance of  being solved if you preprocess the data yourself!

Step 3 cannot be done with today's technology.  The network needs to output text.  Text is discrete.  So unless you implement a policy gradient algorithm (which is hard but probably doable), this step will just be impossible.   I suggest that you request to train a more conventional language model, like the one in https://arxiv.org/abs/1602.02410.   I also recommend that you do not include a critic network.

Likewise, I recommend providing clarity regarding step 4.  The easiest approach is to use Mechanical Turk, which is not free.  One possibility is to create a twitter account.   

For this task, training a language model is a much more promising approach than using the stochastic neural network. 

I recommend removing the requirement to keep track of previous jokes.    To summarize:  the goal should be to train the biggest baddest LSTM on jokes.  

To improve the LSTM and to make it more interesting, it would be nice if for each joke, you could extract a number of keywords that describe the topic of the joke, or better yet, genre, and train the LSTM to maximize the likelihood of the joke given its topic and genre.   If you preprocess the data yourself and put it online someplace, the chance that someone will solve this problem will increase substantially.

**Comment by iaroslav-ai** [comment]
Ok, thanks for informative feedback :) I will make changes accordingly. I think I even saw some relevant datasets around when I was making a background check.

**Comment by ilyasu123** [comment]
Any updates?

**Comment by iaroslav-ai** [comment]
Sorry for the delay. I will push updates in a day or so.

**Comment by ilyasu123** [comment]
Looks decent.  Will edit and merge later today.

**Comment by ilyasu123** [comment]
Right now the problem is to collect a dataset and then to apply char-rnn to it.

To make it more interesting, I'd add:
 - train a big RNN on a large corpus of text (wikipedia or books), then fine-tune it on jokes.  Does the resulting model produce better jokes than an RNN that was trained on jokes only?

 - is there a way to find a large collection of jokes and their topic?   Then, create a dataset where each joke starts with a few words describing its topic.  It's a good idea b/c this way, once you train the model, you'll be able to generate jokes on any topic that you see fit.

 - [Optional:  the language in the request for research should be more formal, but it is something I could do after the merge]

**Comment by iaroslav-ai** [comment]
Yeah, this seems to be like the most realistic thing so far
- right, sort of like retraining imagenet classifier for some more simple tasks, sounds interesting :)
- I did not include this to keep things simple, but I will include this also
- I can give it a try to make it more formal
I will make changes now according to above comment. If this can cause conflicts of any sort, let me know

**Comment by iaroslav-ai** [comment]
Let me know if there are any additional changes you would like me to make.

**Comment by ilyasu123** [comment]
Will merge tomorrow and make a few edits.

**Comment by ilyasu123** [comment]
So I've merged it but then realized that it's not ready yet.  To make the question really good, you need to download the appropriate dataset from r/jokes.  It's a good idea because otherwise everyone who will attempt to solve the problem will have to redo the gruntwork of getting the dataset.  It's important to do because the small datasets that you've linked to are too small for training a reasonable language model.  I know that there are tools for doing so (like https://github.com/NSchrading/redditDataExtractor).  There is also https://www.reddit.com/r/datasets/comments/3bxlg7/i_have_every_publicly_available_reddit_comment/, which allows you to download all of reddit without using a scraper.   

Do that, and the problem will become extremely good.

**Comment by iaroslav-ai** [comment]
I will take a look at it. As I will have the dataset available, I will create a separate merge, where I will either give a link to dataset or maybe torrent if it will be big (but my guess would be that it is no more than 10 Gb when compressed).

**Comment by ilyasu123** [comment]
Sounds good, looking forward!
